### DISCUSSION TITLE
Is it possible to fail validation if an Optional field has value as null in json ?

### DISCUSSION THREAD

**Discussion by rangarb885** [fastapi/fastapi#6560] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Depends, Query, HTTPException
from typing import Optional
from pydantic import BaseModel, Field

app = FastAPI()


class Model(BaseModel):
    a: Optional[int]
    b: Optional[int] = None
    c: Optional[int] = Field(None, nullable=True)

@app.get("/test", response_model=Model)
def foo(m: Model):
    return m
```

### Description

In my case i want to fail the validation when an Optional field has value as **null** ..

I saw the previous issue https://github.com/tiangolo/fastapi/issues/833 and solution provided. But still the FastAPI validation continues without erroring out when the field has value as **null**

Below is the request which i expect to fail on validation as the field `c` is Optional and marked with `nullable=True`
```
{
    "a" : 1,
    "b" : 2,
    "c":  null
 }
```

I m not sure whether this is a problem on pydantic model or fastapi validation.

Is there something i am missing here..?

### Operating System

macOS

### FastAPI Version

0.68.1

### Python Version

Python 3.9.4

**Comment by ml-evs** [comment, marked as answer]
I would suggest that this issue is more appropriate for pydantic than FastAPI, but to give you my quick answer:

1. My understanding is that `nullable` is *NOT* used for validation by pydantic, but will be passed through to the OpenAPI schema by FastAPI (with no other side effects) (look at this [search for `nullable` through the pydantic source](https://github.com/samuelcolvin/pydantic/search?q=nullable)).
1. I would suggest in your case that you would want to indicate in your OpenAPI schema that `nullable` is false, not true.
1. You can add a [pydantic validator](https://pydantic-docs.helpmanual.io/usage/validators/) to raise an error if `c` is `null`. The key difference compared to your code is that I have not provided a default value for the `c` `Field`, e.g.
    ```python
    from pydantic import validator, Field, BaseModel

    class Model(BaseModel):
        a: Optional[int]
        b: Optional[int] = None
        c: Optional[int] = Field(nullable=False)

    @validator("c")
    def check_null(cls, v):
        if v is None:
            raise ValueError("Field 'c' must be an integer, not {v}".)
     
        return v

    ```
    This forbids you from passing `c=None`, but allows you to leave it unset (which I think is what you want).
 

    ```python
    >>> Model(c=None)
    pydantic.error_wrappers.ValidationError: 1 validation error for Model
    c
      Field 'c' must be an integer, not None (type=value_error)
    >>> Model()
    Model(a=None, b=None, c=None)
    >>> print(Model().c)
    None
   ```

Just note that if unset, the value of `c` will still be `None` in an instance of model, which you may or may not want. If you are converting back-and-forth from JSON/pydantic, then you will need to use the `exclude_unset` parameter of `Model().dict(exclude_unset=True)` to prevent further validation failures.

**Comment by rangarb885** [comment]
Thank you @ml-evs . I will use `@validator` .  Adding the code for future reference here.

```
from fastapi import FastAPI, Depends, Query, HTTPException
from typing import Optional
from pydantic import validator, BaseModel, Field

app = FastAPI()

class Model(BaseModel):
    a: Optional[int]
    b: Optional[int] = None
    c: Optional[int] = Field(nullable=False)

    @validator("c")
    def check_null(cls, v):
        if v is None:
            raise ValueError(f"Field 'c' must be an integer, not {v}")
        return v

@app.get("/test", response_model=Model)
def foo(m: Model):
    return m
```

Request JSON:
```
{
    "a" : 1,
    "b" : 2,
    "c": null
 }
```

Response:
```
{
    "detail": [
        {
            "loc": [
                "body",
                "c"
            ],
            "msg": "Field 'c' must be an integer, not None",
            "type": "value_error"
        }
    ]
}
```
